### PR TITLE
Add function to shutdown the server gracefully.

### PR DESCRIPTION
**Pull request by Viatorus**
Currently, it is not possible to shut down the aiohttp server gracefully, because the _service_task is running in an endless loop.

This PR should fix this problem.

### CODE REVIEW

File: engineio/asyncio_server.py [hunk 1]
```diff
@@ -220,24 +231,37 @@ def start_background_task(self, target, *args, **kwargs):
...
+        if not until_shutdown:
+            return await asyncio.sleep(seconds)
+
+        # Sleep a short time and recheck if server is still running.
+        fractional, integer = math.modf(seconds * 10)
+        for i in range(int(integer)):
+            await asyncio.sleep(0.1)
+            if not self.running:
+                return
+        return await asyncio.sleep(fractional)
```
Comment (miguelgrinberg) [inline engineio/asyncio_server.py]: What you are doing here is a somewhat busy wait loop. This is going to add CPU load, so it's not something that I find acceptable. I would actually prefer to leave the sleep function working as a sleep function, and build a better mechanism for the task service to be awaken if it needs to exit, maybe with an `Event` object or similar.

**Comment by miguelgrinberg** [comment]
@Viatorus Thanks! Overall I'm okay with the idea of adding the `shutdown()` method.

The changes to the `sleep()` function that you made are not the way I would do it though, because you rely on polling to determine when it is time to exit and that requires constant CPU work. Let me know if you want to continue working on the PR or not, I can do this myself as well. Thanks again.

**Comment by Viatorus** [comment]
@miguelgrinberg Thanks for your feedback. I updated this part. This change seems to look much cleaner to me. What do you think?

I have never used async/await before.

File: engineio/asyncio_server.py [hunk 2]
```diff
@@ -63,6 +67,13 @@ def attach(self, app, engineio_path='engine.io'):
         engineio_path = engineio_path.strip('/')
         self._async['create_route'](app, self, '/{}/'.format(engineio_path))
 
+    async def shutdown(self):
+        """Shutdown the Engine.IO server."""
+        self.running = False
```
Comment (miguelgrinberg) [inline engineio/asyncio_server.py]: You do not need `self.running` anymore, right?

File: engineio/asyncio_server.py [hunk 3]
```diff
@@ -63,6 +67,13 @@ def attach(self, app, engineio_path='engine.io'):
...
+
+        self.__service_task.cancel()
+        await self.__service_task
```
Comment (miguelgrinberg) [inline engineio/asyncio_server.py]: How long does this take? It would seem you could be waiting up to a full minute until the task exits.

**Comment by Viatorus** [comment]
Thank you for your feedback.

It didn't worked this way, because a task never reaches the state cancelled if it is running. There was also an mistake with: `while self.__service_task.cancelled()`.

Now it should work. The cancel exception raises immediately (also while sleeping). This is now the break condition. No waiting required.

**Comment by miguelgrinberg** [comment]
Thanks, this looks better. I need to test it a bit, and also add a compatible feature for the synchronous server case. As soon as I have that worked out I'll merge and make a new release.

**Comment by Viatorus** [comment]
Great, thank you!

**Comment by miguelgrinberg** [comment]
@Viatorus how are you shutting down aiohttp gracefully? I don't really see what the problem is, when I hit Ctrl-C to stop the server nothing happens, and I need to hit Ctrl-C a second time to get the server to exit. This happens even if I don't start the service task.

**Comment by Viatorus** [comment]
Did you connect to your socket.io first?

**Comment by miguelgrinberg** [comment]
@Viatorus I'm using the `examples/server/aiohttp/latency.py` example in this repository.

**Pull request closed without merging**